data/generate_embeddings.py

```
@@ -0,0 +1,598 @@
+    
+    return text
+
+
+def discover_input_files(input_dir: Path) -> List[Path]:
+    """Find all supported input files recursively."""
+    all_files: List[Path] = []
+    
+    for root, _, files in os.walk(input_dir):
+        for f in files:
+            p = Path(root) / f
+            if p.suffix.lower() in SUPPORTED_EXTENSIONS:
+                all_files.append(p)
+    
+    return sorted(all_files)
+
+
+def load_metadata_mappings(csv_path: Path) -> Tuple[Dict[str, str], Dict[str, str]]:
+    """Load metadata CSV and build mappings for city lookup.
+
+    Returns:
+      - path_to_city: maps normalized posix file path (e.g., 'data/raw/Kyiv_....txt') -> city
+      - basename_to_city: maps basename (e.g., 'Kyiv_....txt') -> city if unique, otherwise omitted
+    """
+    path_to_city: Dict[str, str] = {}
+    basename_counts: Dict[str, int] = {}
+    basename_temp: Dict[str, str] = {}
+
+    if not csv_path.exists():
+        print(f"Warning: metadata CSV not found at {csv_path}")
+        return path_to_city, {}
+
+    with csv_path.open("r", encoding="utf-8") as f:
+        reader = csv.DictReader(f)
+        for row in reader:
+            city = (row.get("city") or "").strip()
+            file_path = (row.get("file_path") or "").strip()
+            if not city or not file_path:
+                continue
+            rel_path = Path(file_path)
+            key_rel = rel_path.as_posix()
+            path_to_city[key_rel] = city
+            # Also map absolute path resolved from project root (cwd)
+            if not rel_path.is_absolute():
+                abs_key = (Path.cwd() / rel_path).resolve().as_posix()
+                path_to_city[abs_key] = city
+            base = rel_path.name
+            basename_counts[base] = basename_counts.get(base, 0) + 1
+            basename_temp[base] = city
+
+    basename_to_city: Dict[str, str] = {b: basename_temp[b] for b, c in basename_counts.items() if c == 1}
+    return path_to_city, basename_to_city
```

Validate that CSV columns exist and have the right headers, failing fast if not.
Print a summary of how many mappings were loaded for debugging.

---

Thanks, done!